pkg/operator/ceph/cluster/controller.go

```diff
@@ -261,58 +273,119 @@ func (r *ReconcileCephCluster) reconcile(request reconcile.Request) (reconcile.R
+		}
+		return opcontroller.WaitForRequeueIfFinalizerBlocked, cephCluster, errors.New(dependentMsg)
+	}
+
+	deletingMsg := fmt.Sprintf("deleting CephCluster %q", cephCluster.Name)
+	// Report an event that the cluster is deleting
+	r.clusterController.recorder.ReportIfNotPresent(cephCluster, corev1.EventTypeNormal, string(cephv1.ClusterDeletingReason), deletingMsg)
+	// Set the DeletionIsBlocked=False status condition
+	cond := buildDeletionBlockedDueToDependentsCondition(false, "cluster can be deleted safely")
+	cephv1.SetStatusCondition(&cephCluster.Status.Conditions, cond)
+	if err := opcontroller.UpdateStatus(r.client, cephCluster); err != nil {
+		logger.Warningf("failed to set CephCluster %q status condition %s=%s; continuing deletion without setting the condition", r.clusterController.namespacedName.String(), cond.Type, cond.Status)
+	}
+	if cephCluster, err = r.refreshCephCluster(cephCluster); err != nil {
+		return reconcile.Result{}, cephCluster, err
+	}
+	// Log deletion to the operator log
+	logger.Info(deletingMsg)
+
+	doCleanup := true
+
+	// Start cluster clean up only if cleanupPolicy is applied to the ceph cluster
+	stopCleanupCh := make(chan struct{})
+	if cephCluster.Spec.CleanupPolicy.HasDataDirCleanPolicy() && !cephCluster.Spec.External.Enable {
+		monSecret, clusterFSID, err := r.clusterController.getCleanUpDetails(cephCluster.Namespace)
+		if err != nil {
+			logger.Warningf("failed to get mon secret. Skip cluster cleanup and remove finalizer. %v", err)
+			doCleanup = false
 		}
 
 		if doCleanup {
-			// Run delete sequence
-			response, ok := r.clusterController.requestClusterDelete(cephCluster)
-			if !ok {
-				// If the cluster cannot be deleted, requeue the request for deletion to see if the conditions
-				// will eventually be satisfied such as the volumes being removed
-				close(stopCleanupCh)
-				return response, cephCluster, nil
+			cephHosts, err := r.clusterController.getCephHosts(cephCluster.Namespace)
+			if err != nil {
+				return reconcile.Result{}, cephCluster, errors.Wrapf(err, "failed to find valid ceph hosts in the cluster %q", cephCluster.Namespace)
 			}
+			go r.clusterController.startClusterCleanUp(stopCleanupCh, cephCluster, cephHosts, monSecret, clusterFSID)
 		}
+	}
 
-		// Remove finalizer
-		err = removeFinalizer(r.client, request.NamespacedName)
+	if doCleanup {
+		// Run delete sequence
+		response, err := r.clusterController.requestClusterDelete(cephCluster)
 		if err != nil {
-			return reconcile.Result{}, cephCluster, errors.Wrap(err, "failed to remove finalizer")
+			// If the cluster cannot be deleted, requeue the request for deletion to see if the conditions
+			// will eventually be satisfied such as the volumes being removed
+			// TODO: ask Travis/Sev - do we never clean up the stopCleanupCh if this is successful?
+			//       Is this a memory leak? why is this in a goroutine in the first place?
+			close(stopCleanupCh)
```

Comment:
But what if the jobs weren't started yet and end up not being created because the channel is closed at completion?

AFAICT, this doesn't wait on the jobs to finish -- only on the jobs to start.

Reply:
I see now, you're right we don't always want to close the channel. It should only be closed here when the call to `requestClusterDelete()` fails. If `requestClusterDelete()` succeeds, we want to proceed and remove the finalizer, which will allow the ceph daemons to be deleted, which will allow the goroutine to start the job. So the goroutine should close the channel when it's done.

Reply:
@sp98 may have more insight as well